changelog.d/9247.feature

```diff
@@ -0,0 +1 @@
+Add experimental support for backfilling history into rooms ([MSC2716](https://github.com/matrix-org/matrix-doc/pull/2716)).
```

Comment:
@erikjohnston Can you help with reviewing the actual code here so we can move closer to merging this to `master`? 

 - Pointers around my `TODO` questions
 - Where I should be organizing things away to to dry up some of the code
 - Any tests I should be adding to Synapse itself?
 - Feature flag: The `/bulksend` endpoint is already behind a feature flag. But there are some changes to the internals of Synapse as well, should we also feature flag those for extra safety?

Reply:
Yup, will do!